Add c make project export

hjmjohnson opened the pull request.

hjmjohnson commented:
@wjakob This contains main improvements for the cmake build system.  I'll watch the CI dashboards to see if it succeeds in building.

The new compiler flags are added for when clang compiler on linux is available (with override).

Major improvement to provide cmake package support internally (No module needed) to facilitate linking into external applications.

hjmjohnson commented:
@wjakob I have cleaned up the initial build errors related to making the build process backwards compatible to 2.8.7.

It appears that the current windows builds are not working, but I need to work on grant writing today instead of windows debugging.

Hans

hjmjohnson commented:
@wjakob Well.... after a lot of testing and re-arranging to get all the pieces into place for the windows builds, and the cmake 2.8.7 version on linux, I think this is finally ready for merging!

Please review and consider these changes as a way to make building against TBB easier from external packages.

Regards,
Hans

hjmjohnson commented:
@wjakob NOTE: This pulll request superceeds the other one, so there is no need to do both.

wjakob commented on CMakeLists.txt:
```diff
@@ -1,4 +1,59 @@
+    set(CMAKE_CXX_STANDARD "11") ## Default to cmake standard version 11 if not previously set
+  endif()
+endif()
+message(STATUS "Building with CMAKE_CXX_STANDARD=${CMAKE_CXX_STANDARD}")
```
I'd like to keep the build as quiet as possible since this is included in many other projects.

hjmjohnson replied:
Done

wjakob commented on CMakeLists.txt:
```diff
@@ -1,4 +1,59 @@
+# CMake buld configuration for tbb
+#
+if(DEFINED CMAKE_CXX_STANDARD AND (NOT ${CMAKE_CXX_STANDARD} EQUAL "98" ))
+  cmake_minimum_required(VERSION 3.2.0 FATAL_ERROR) ## 3.2.0 needed for AppleClang CXX11 finding
```
"finding" -> "detection"

wjakob commented on CMakeLists.txt:
```diff
@@ -1,4 +1,59 @@
+    check_cxx_compiler_flag ("-std=c++11" SUPPORTS_STDCXX11)
+    if (SUPPORTS_STDCXX11)
+      set (CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
+      set(CMAKE_CXX_STANDARD "11") ## Default to cmake standard version 11 if not previously set
```
can remove this

hjmjohnson replied:
This could be removed, but that puts the burden back on end-users to know distinguish between various compiler nuances that the cmake team has already addressed.   

A more detailed explaination of features that are available in newer versions of cmake are:
https://crascit.com/2015/03/28/enabling-cxx11-in-cmake/

I've been working with the cmake developers on incorporating these features in other large heavily used projects (www.itk.org, www.vtk.org, http://vxl.sourceforge.net).

wjakob commented on CMakeLists.txt:
```diff
@@ -1,4 +1,59 @@
+      set (CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
+      set(CMAKE_CXX_STANDARD "11") ## Default to cmake standard version 11 if not previously set
+    else()
+      set(CMAKE_CXX_STANDARD "98") ## Default to cmake standard version 98 if -std=c++11 not supported
```
can remove this

hjmjohnson replied:
Being complete here.  Explicitly providing the intended CXX_STANDARD that will be compiled against.

wjakob commented on CMakeLists.txt:
```diff
@@ -128,6 +133,7 @@ if (UNIX)
   check_cxx_compiler_flag ("-mrtm" SUPPORTS_MRTM)
   if (SUPPORTS_MRTM)
     set (CMAKE_CXX_FLAGS "-mrtm ${CMAKE_CXX_FLAGS}")
+    message(STATUS "NOTE: Using -mrtm flag for compiler")
```
Please remove to keep build quiet.

hjmjohnson replied:
Done

wjakob commented on CMakeLists.txt:
```diff
@@ -86,7 +86,12 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
 
-include_directories(include src src/rml/include ${CMAKE_CURRENT_BINARY_DIR})
+
+include_directories(${CMAKE_CURRENT_SOURCE_DIR}/src
```
Is this actually needed? I think CMake compilation commands are always relative to ${CMAKE_CURRENT_SOURCE_DIR}

hjmjohnson replied:
Sure.  I just prefer being explicit rather than depending on implicit rules.

wjakob commented on CMakeLists.txt:
```diff
@@ -352,7 +357,7 @@ else ()
 endif()
 include_directories (${CMAKE_BINARY_DIR})
-configure_file (build/version_string.ver.in version_string.ver @ONLY)
+configure_file (${CMAKE_CURRENT_LIST_DIR}/build/version_string.ver.in version_string.ver @ONLY)
```
This I think should be ${CMAKE_CURRENT_BINARY_DIR}, if anything. CMAKE_CURRENT_LIST_DIR will clutter up the input tree with an extra file during an out-of-tree build.

hjmjohnson replied:
This will not clutter the input:
This is just being explicit on the input value 
configure_file(<input> <output>

The above code does not change anything.  I'll revert to use the implicit pathing like it used to be.

wjakob commented on CMakeLists.txt:
```diff
@@ -86,6 +86,10 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
     "MinSizeRel" "RelWithDebInfo")
 endif()
 
```
Regarding this overall commit: can you clarify how package support helps including TBB in other cmake based packages?

The standard way this project is generally used is via a simple ``add_subdirectory`` command. This ensures that TBB is built along with the main program, and that everything is compiled in the right order to satisfy dependencies. I have never observed non-robust behavior on any platform (linux, macos, windows).

hjmjohnson replied:
CMake provides a find_package() mechansim for including external dependancies.  The typical scenerio I have encountered is
1) download, compile, install package (tbb in this case)
2) from the many tools that depnd on TBB, use find_package to point to that one build of tbb


Then from independant build add the following to it's CMakeList.txt

find_package(tbb REQUIRED)
add_executable(test_tbb test.cpp)
target_link_libraries(test_tbb PUBLIC tbb::tbb )

hjmjohnson replied:
PS:  This is the mechansims used by LAPACK, ITK, VTK, QT, and many many other cmake based packages.

wjakob commented on CMakeLists.txt:
```diff
@@ -86,6 +86,22 @@ if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)
+endif()
+SETIFEMPTY(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/lib)
+SETIFEMPTY(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/lib)
+SETIFEMPTY(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/bin)
```
What bout windows? It uses release/debug/.. versions of all of these variables.

hjmjohnson replied:
This is similar to the mechanism to set these variables for ITK, vxl, and VTK that are extensively tested on windows.

See dashboard at: https://open.cdash.org/index.php?project=Insight&date=2016-09-21

I think this works just fine with all the various windows testing plafortforms for those packages.

hjmjohnson commented:
@wjakob This should now be ready for merging.  I've addressed all of the issues that you have presented.

I do plan to submit this build system upstream to the main TBB development with the hopes that it will be included in in future releases.

hjmjohnson commented:
I am breaking this appart into smaller pull requests.
